File: RipInfo.py
from __future__ import annotations

import os
import string
from typing import Iterator

import google.auth.exceptions
import googleapiclient.errors
from getfilelistpy import getfilelist
from google.auth.transport.requests import Request
from google.oauth2.credentials import Credentials
from google_auth_oauthlib.flow import InstalledAppFlow

from Enums import FilenameScheme
from ImageLink import ImageLink

SCOPES = ['https://www.googleapis.com/auth/drive.readonly']


class RipInfo:
    """Ripped Site Information"""

    translation_table = dict.fromkeys(map(ord, '<>:"/\\|?*.'), None)
    gdrive_creds: Credentials = None

    def __init__(self, urls: list[str | ImageLink] | str, dir_name: str = "",
                 filename_scheme: FilenameScheme = FilenameScheme.ORIGINAL,
                 generate: bool = False, num_urls: int = 0, filenames: list[str] = None):
        if isinstance(urls, str) or isinstance(urls, ImageLink):
            urls = [urls]
        self.filename_scheme: FilenameScheme = filename_scheme
        self.__dir_name: str = dir_name
        self.__filenames: list[str] = filenames
        if self.__dir_name:
            self.__dir_name = self.__clean_dir_name(self.__dir_name)
        self.urls: list[ImageLink] = self.__convert_urls_to_image_link(urls)
        self.must_generate_manually: bool = generate
        self.url_count = num_urls if generate else len(self.urls)

    def __str__(self) -> str:
        return f"({[str(url) for url in self.urls]}, {self.num_urls}, {self.dir_name})"

    def __iter__(self) -> Iterator[ImageLink]:
        return iter(self.urls)

    @property
    def num_urls(self):
        return self.url_count

    @property
    def dir_name(self):
        return self.__dir_name

    @dir_name.setter
    def dir_name(self, value):
        self.__dir_name = value
        self.__dir_name = self.__clean_dir_name(self.__dir_name)

    def __convert_urls_to_image_link(self, urls: list[str | ImageLink]) -> list[ImageLink]:
        image_links = []
        link_counter = 0
        filename_counter = 0
        urls = self.__remove_duplicates(urls)
        for url in urls:
            if isinstance(url, ImageLink):
                image_links.append(url)
                continue
            if "drive.google.com" in url:
                try:
                    image_link, link_counter = self.__query_gdrive_links(url, link_counter)
                    image_links.extend(image_link)
                except googleapiclient.errors.HttpError:
                    pass
            else:
                filename = self.__filenames[filename_counter] if self.__filenames else ""
                filename_counter += 1
                image_link = ImageLink(url, self.filename_scheme, link_counter, filename=filename)
                image_links.append(image_link)
                link_counter += 1
        return image_links

    @staticmethod
    def __remove_duplicates(urls: list[str | ImageLink]) -> list[str | ImageLink]:
        unique_urls: list[str | ImageLink] = []
        seen_urls = set()
        for url in urls:
            if isinstance(url, ImageLink):
                link = url.url
            else:
                link = url

            if link in seen_urls:
                continue
            else:
                unique_urls.append(url)
                seen_urls.add(link)
        return unique_urls

    def __query_gdrive_links(self, gdrive_url: str, index: int) -> tuple[list[ImageLink], int]:
        self.authenticate()
        id_, single_file = self.__extract_id(gdrive_url)
        resource = {
            "id": id_,
            "oauth2": self.gdrive_creds,
            "fields": "files(name,id)",
        }
        res = getfilelist.GetFileList(resource)
        if not self.__dir_name:
            dir_name = res["searchedFolder"]["name"] if not single_file else res["searchedFolder"]["id"]
            self.__dir_name = self.__clean_dir_name(dir_name)
        links: list[ImageLink] = []
        counter = index
        if single_file:
            filename = res["searchedFolder"]["name"]
            file_id = id_
            img_link = ImageLink(file_id, self.filename_scheme, counter, filename=filename, gdrive=True)
            links.append(img_link)
            counter += 1
        else:
            file_lists = res["fileList"]
            folder_ids = res["folderTree"]["id"]
            folder_names = res["folderTree"]["names"]
            folder_names = self.get_gdrive_folder_names(folder_ids, folder_names)
            for i, file_list in enumerate(file_lists):
                files = file_list["files"]
                parent_folder = folder_names[i]
                for file in files:
                    file_id = file["id"]
                    filename = os.path.join(parent_folder, file["name"])
                    img_link = ImageLink(file_id, self.filename_scheme, counter, filename=filename, gdrive=True)
                    links.append(img_link)
                    counter += 1
        return links, counter

    def get_gdrive_folder_names(self, folder_ids: list[list[str]], names: list[str]) -> list[str]:
        # Stores the mapping from folder id to folder name
        hierarchy = {}
        for i, name in enumerate(names):
            folder_id = folder_ids[i]
            complete = False
            for id_ in folder_id:
                parent = hierarchy.get(id_, None)
                if not parent:
                    hierarchy[id_] = self.__clean_dir_name(name)
                    complete = True
                    break
            if complete:
                continue

        # Using the id -> name mapping, resolve the full path of each file
        folder_names = []
        for id_set in folder_ids:
            path_ = ""
            for id_ in id_set:
                if not path_:
                    path_ = hierarchy[id_]
                else:
                    path_ = os.path.join(path_, hierarchy[id_])
            folder_names.append(path_)
        return folder_names

    @staticmethod
    def __extract_id(url: str) -> tuple[str, bool]:
        parts = url.split("/")
        if "/d/" in url:
            return parts[-2], True
        else:
            id_ = parts[-1].split('?')[0]
            if id_ in ("open", "folderview"):
                id_ = parts[-1].split('?id=')[-1]
            return id_, False

    @staticmethod
    def authenticate():
        if os.path.exists('token.json'):
            RipInfo.gdrive_creds = Credentials.from_authorized_user_file('token.json', SCOPES)

        # If there are no (valid) credentials available, let the user log in.
        if not RipInfo.gdrive_creds or not RipInfo.gdrive_creds.valid:
            if RipInfo.gdrive_creds and RipInfo.gdrive_creds.expired and RipInfo.gdrive_creds.refresh_token:
                try:
                    RipInfo.gdrive_creds.refresh(Request())
                except google.auth.exceptions.RefreshError:
                    flow = InstalledAppFlow.from_client_secrets_file("credentials.json", SCOPES)
                    RipInfo.gdrive_creds = flow.run_local_server(port=0)
            else:
                flow = InstalledAppFlow.from_client_secrets_file("credentials.json", SCOPES)
                RipInfo.gdrive_creds = flow.run_local_server(port=0)

            # Save the credentials for the next run
            with open("token.json", "w") as token:
                token.write(RipInfo.gdrive_creds.to_json())

    def __clean_dir_name(self, dir_name: str) -> str:
        """
            Remove forbidden characters from path
        """
        dir_name = dir_name.translate(self.translation_table).strip().replace("\n", "")
        if dir_name[-1] not in (")", "]", "}"):
            dir_name.rstrip(string.punctuation)
        if dir_name[0] not in ("(", "[", "{"):
            dir_name.lstrip(string.punctuation)
        return dir_name

    @classmethod
    def deserialize(cls, object_data: dict) -> RipInfo:
        rip_info = cls([], "")
        rip_info.urls = [ImageLink.deserialize(data) for data in object_data["urls"]]
        rip_info.url_count = object_data["url_count"]
        rip_info.dir_name = object_data["dir_name"]
        rip_info.must_generate_manually = object_data["must_generate_manually"]
        return rip_info

    def serialize(self) -> dict[str, any]:
        object_data = {
            "urls": [img_lnk.serialize() for img_lnk in self.urls],
            "url_count": self.url_count,
            "dir_name": self.dir_name,
            "must_generate_manually": self.must_generate_manually
        }
        return object_data


if __name__ == "__main__":
    pass

File: Core/RipInfo.cs
using System.Text;
using Core.Enums;
using Google;

namespace Core;

public class RipInfo
{
    private static readonly HashSet<char> ForbiddenChars = ['<', '>', ':', '"', '/', '\\', '|', '?', '*'];
        
    private string _directoryName = null!; // Initialized through the property setter
    
    public FilenameScheme FilenameScheme { get; set; }
    public List<ImageLink> Urls { get; set; }
    public bool MustGenerateManually { get; set; }
    
    private int UrlCount { get; set; }

    public int NumUrls => UrlCount;

    public string DirectoryName
    {
        get => _directoryName;
        set => _directoryName = ClearDirectoryName(value);
    }

    private List<string>? Filenames { get; set; }

    public RipInfo(List<StringImageLinkWrapper> urls, string directoryName = "", FilenameScheme filenameScheme = FilenameScheme.Original,
        bool generate = false, int numUrls = 0, List<string>? filenames = null, bool discardBlobs = false)
    {
        // SaveRawUrls(urls);
        FilenameScheme = filenameScheme;
        DirectoryName = directoryName;
        Filenames = filenames;
        Urls = ConvertUrlsToImageLink(urls, discardBlobs).Result;
        MustGenerateManually = generate;
        UrlCount = generate ? numUrls : Urls.Count;
    }

    private async Task<List<ImageLink>> ConvertUrlsToImageLink(List<StringImageLinkWrapper> urls, bool discardBlob)
    {
        var imageLinks = new List<ImageLink>();
        var linkCounter = 0;        // Current index of image_links (used for naming image_links when generating numeric names)
        var filenameCounter = 0;    // Current index of filenames
        urls = RemoveDuplicates(urls);
        foreach (var url in urls)
        {
            if (url.ImageLink is not null)
            {
                if (FilenameScheme == FilenameScheme.Chronological)
                {
                    url.ImageLink.Rename(linkCounter);
                }
                linkCounter++;
                imageLinks.Add(url.ImageLink);
                continue;
            }

            if (url.Url is null)
            {
                continue;
            }

            if (url.Url.Contains("drive.google.com"))
            {
                try
                {
                    var (imageLink, newLinkCounter) = await QueryGDriveLinks(url.Url, linkCounter);
                    imageLinks.AddRange(imageLink);
                    linkCounter = newLinkCounter;
                }
                catch (GoogleApiException _) // googleapiclient.errors.HttpError
                {
                    // pass
                }
            }
            else
            {
                var filename = Filenames?[filenameCounter] ?? "";
                filenameCounter++;
                var imageLink = new ImageLink(url.Url, FilenameScheme, linkCounter, filename: filename);
                imageLinks.Add(imageLink);
                linkCounter++;
            }
        }
        if (discardBlob)
        {
            imageLinks = imageLinks.Where(imageLink => !imageLink.IsBlob).ToList();
        }
        return imageLinks;
    }

    private async Task<(List<ImageLink>, int)> QueryGDriveLinks(string gDriveUrl, int index)
    {
        var service = await GDriveHelper.AuthenticateGDrive();
        var (id, singleFile) = ExtractId(gDriveUrl);
        var files = await service.GetFiles(id);
        var imageLinks = new List<ImageLink>();
        var counter = index;
        foreach (var file in files)
        {
            var imgLink = new ImageLink(file.Id, FilenameScheme, counter, filename: file.Name, linkInfo: LinkInfo.GDrive);
            imageLinks.Add(imgLink);
            counter++;
        }
        return (imageLinks, counter);
    }

    private static (string, bool) ExtractId(string url)
    {
        var parts = url.Split("/");
        if (url.Contains("/d/"))
        {
            return (parts[^2], true);
        }

        var id = parts[^1].Split('?')[0];
        if (id is "open" or "folderview")
        {
            id = parts[^1].Split("?id=")[^1];
        }
        return (id, false);
    }
    
    private static List<StringImageLinkWrapper> RemoveDuplicates(List<StringImageLinkWrapper> urls)
    {
        var urlSet = new HashSet<string>();
        var newUrls = new List<StringImageLinkWrapper>();
        foreach (var url in urls)
        {
            if (url.Url is not null && urlSet.Add(url.Url))
            {
                newUrls.Add(url);
            }
            else if (url.ImageLink is not null && urlSet.Add(url.ImageLink.Url))
            {
                newUrls.Add(url);
            }
        }
        return newUrls;
    }
    
    private static string ClearDirectoryName(string directoryName)
    {
        var dirName = new StringBuilder();
        foreach (var c in directoryName)
        {
            if (!ForbiddenChars.Contains(c))
            {
                dirName.Append(c);
            }
        }
        if (dirName[^1] != ')' && dirName[^1] != ']' && dirName[^1] != '}')
        {
            dirName.Remove(dirName.Length - 1, 1);
        }
        if (dirName[0] != '(' && dirName[0] != '[' && dirName[0] != '{')
        {
            dirName.Remove(0, 1);
        }
        return dirName.ToString();
    }
    
    private static void SaveRawUrls(List<StringImageLinkWrapper> urls)
    {
        JsonUtility.Serialize("raw_urls.json", urls);
    }

    public override string ToString()
    {
        return $"([{string.Join(", ", Urls.Select(url => url.ToString()))}], {NumUrls}, {DirectoryName})";
    }
}